### ISSUE TITLE
Question about numGradients in setGradients()

### ISSUE THREAD

**Issue by bgyoon** [facebookresearch/moolib#48]
Why is gradients in `setGradients()` [divided by data.numGradients](https://github.com/facebookresearch/moolib/blob/337faa9ddacda619a720a29c101afe80abef7dc3/src/accumulator.cc#L432)? Shouldn't it divided by data.batchSize?

            grad.mul_(1.0f / data.numGradients);

`data.numGradients `seems to be the number of `reduce_gradients()` called. `data.batchSize` is sum of `batch_size` passed in `reduce_gradients()`. `numGradients` doesn't seem to be a good indicator of how many iterations worth of gradients are accumulated.

For example,
- set_virtual_batch_size(8)
- reduce_gradients(batch_size=4) is called -> numGradients += 1
- reduce_gradients(batch_size=3) is called -> numGradients += 1
- reduce_gradients(batch_size=2) is called -> numGradients += 1
- (after gradients all-reduce) data.batchSize=9, data.numGradients=3

Shouldn't gradients be divided by 9, not 3?

**Comment by tscmoo** [comment]
You're right it makes more sense in some ways, however, usually people use the same batch_size for all peers, in which case diving by numGradients allows the magnitude of the gradients to remain unchanged when training on a single node vs many nodes.
If we do as you suggest (divide by the virtual batch size), it presents a few problems:
1. if the user divides the loss by the (local) batch size before calling reduce_gradients, we would *also* divide it by the virtual batch size, which is obviously incorrect.
2. if the user doesn't divide the loss, we now effectively force it to be divided by the virtual batch size, which may be unexpected.

Dividing by num gradients instead works more like what the user might expect. eg. if they divide the loss, it stays divided (equivalent to dividing by virtual batch size if all peers use the same batch size), and if they don't divide the loss, then the magnitude of the loss is not dependent of the number of peers that participate.

Ideally this would be configurable such that the user could choose in which matter the gradients are reduced, though.

**Comment by bgyoon** [comment]
Ahhh, I see. `reduce_gradients()` is expected to work like a reduce mean, where sum of gradients is divided by number of `reduce_gradients()` call. That makes sense.


Just FYI, my use case was training a transformer model using moolib. reduce_gradients() seems to incur GPU->CPU copy even when all-reduce isn't performed(i.e., `*size < virtualBatchSize`) I hacked around it by letting each peer accumulate gradients until enough global batches have been processed. I manually checked the processed global batch size by calling `group.all_reduce(local_batch_size)` in my Python code. I would only call `reduce_gradients()` when only I know all-reduce is going to happen, to ensure single GPU->CPU call per gradients all-reduce. Since each peer may accumulate gradients at different rate(mainly due to rpc, all-reduce delay, even though they're of the same hardware spec), I found `reduce_gradients()` being called with different batch_size by each peers. This isn't really an issue for me, just FYI...
